R/drug_response_core.R

```diff
@@ -0,0 +1,463 @@
+  }
+  if ((!is.null(remove_outliers_sd) || !is.null(winsorize_pct)) && !is.null(range_sd_filter)) {
+    stop("`range_sd_filter` cannot be used with `remove_outliers_sd` or `winsorize_pct`.")
+  }
+
+  # 1. Get all relevant lab measurements and drug purchases
+  lab_measurements <- get_lab_measurements(conn$labs, lablist, require_values = TRUE,
+                                           covariates = covariates, covariate_cols = covariate_cols)
+
+  first_purchases <- get_first_purchase(conn$pheno, druglist) %>%
+    select(FINNGENID, first_drug_age = EVENT_AGE)
+
+  # 2. Join lab data with first purchase data
+  all_measurements <- left_join(lab_measurements, first_purchases, by = "FINNGENID") %>%
+    mutate(time_to_drug = .data$first_drug_age - .data$EVENT_AGE)
+
+  # 3. Handle outlier removal
+  if (!is.null(range_sd_filter)) {
+    # Validate the filter parameters
+    required_params <- c("lower_bound", "upper_bound", "nsd")
+    if (!is.list(range_sd_filter) || !all(required_params %in% names(range_sd_filter))) {
+      stop("`range_sd_filter` must be a list containing `lower_bound`, `upper_bound`, and `nsd`.")
+    }
+
+    # Calculate mean and sd on values within the specified range
+    ranged_data <- all_measurements %>%
+      filter(.data$MEASUREMENT_VALUE_HARMONIZED >= range_sd_filter$lower_bound &
+             .data$MEASUREMENT_VALUE_HARMONIZED <= range_sd_filter$upper_bound)
+
+    mean_val <- mean(ranged_data$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+    sd_val <- sd(ranged_data$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+
+    # Define outlier bounds based on the ranged statistics
+    lower_bound_stat <- mean_val - (range_sd_filter$nsd * sd_val)
+    upper_bound_stat <- mean_val + (range_sd_filter$nsd * sd_val)
+
+    original_rows <- nrow(all_measurements)
+    all_measurements <- all_measurements %>%
+      filter(.data$MEASUREMENT_VALUE_HARMONIZED >= lower_bound_stat & .data$MEASUREMENT_VALUE_HARMONIZED <= upper_bound_stat)
+    removed_rows <- original_rows - nrow(all_measurements)
+    print(paste("Removed", removed_rows, "outliers using range_sd_filter."))
+  }
+
+  if (!is.null(remove_outliers_sd)) {
+    mean_val <- mean(all_measurements$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+    sd_val <- sd(all_measurements$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+    lower_bound <- mean_val - (remove_outliers_sd * sd_val)
+    upper_bound <- mean_val + (remove_outliers_sd * sd_val)
+
+    all_measurements <- all_measurements %>%
+      filter(.data$MEASUREMENT_VALUE_HARMONIZED >= lower_bound & .data$MEASUREMENT_VALUE_HARMONIZED <= upper_bound)
+  }
+
+  if (!is.null(winsorize_pct)) {
+    all_measurements <- all_measurements %>%
+      mutate(MEASUREMENT_VALUE_HARMONIZED = winsorize_vector(.data$MEASUREMENT_VALUE_HARMONIZED, winsorize_pct))
+  }
+
+  # 4. Filter measurements based on the time window for exposed individuals
+  time_window_years <- months_before / 12
+
+  exposed_measurements <- all_measurements %>%
```

Comment:
could add the unexposed_measurements condition also here so that those without time to drug are included so no need to filter twice and bind_rows